cpp/src/io/orc/writer_impl.hpp

```diff
@@ -330,6 +332,24 @@ class writer::impl {
   bool enable_dictionary_ = true;
   bool enable_statistics_ = true;
 
+  /// Overall file metadata.  Filled in during the process and written during write_chunked_end()
+  cudf::io::orc::FileFooter ff;
+  cudf::io::orc::Metadata md;
+  /// current write position for rowgroups/chunks
+  size_t current_chunk_offset;
+  /// optional user metadata
+  table_metadata const* user_metadata = nullptr;
+  /// only used in the write_chunked() case. copied from the (optionally) user supplied
+  /// argument to write_chunked_begin()
+  table_metadata_with_nullability user_metadata_with_nullability;
+  /// special parameter only used by detail::write() to indicate that we are guaranteeing
+  /// a single table write.  this enables some internal optimizations.
+  bool const single_write_mode;
+  /// To track whether necessary members have been initialized
+  bool initialized = false;
+  /// to track if the output has been written to sink
+  bool closed = false;
```

Comment:
Maybe this could be combined into a single state, given that not all combinations of flags are valid. I'll look more into this part tomorrow.